### PR TITLE
Union and improve size field functions

### PR DESCRIPTION
**Pull request by BinWang0213**
both gmsh occ and geo should support size field. 

In this pull request:
1. Create a separate module file `size_field.py` to support various size field in gmsh
2. Add `gmsh.model.mesh.field.setNumber(tag1, "NumPointsPerCurve", 100)`  in BoundaryLayer to avoid insufficient refinement over edges
3. Set up `ignore_other_mesh_sizes` whenever we applied `SetBackgroundMesh`. This is the typical usage to avoid overrefinement in gmsh

I did few testings which work well.

Please let me know if this looks good to merge.
Bin

### CODE REVIEW

**Comment by nschloe** [comment]
> Create a separate module file size_field.py to support various size field in gmsh

Ok.

> Add gmsh.model.mesh.field.setNumber(tag1, "NumPointsPerCurve", 100) in BoundaryLayer to avoid insufficient refinement over edges

This is too arbitrary. Please remove. You can add a setter + getter like for the other values.

> Set up ignore_other_mesh_sizes whenever we applied SetBackgroundMes

Please remove. One can already set those values to 0 if so desired.

Also, this needs a small test. Perhaps you can move the boundary refinement test into /test/ and add one for occ.

**Comment by BinWang0213** [comment]
Thanks for your comments! I did changes based on your suggestions.
Please review the code when you are available!

File: pygmsh/occ/geometry.py [hunk 1]
```diff
@@ -55,6 +56,15 @@ def revolve(self, *args, **kwargs):
         assert angle < 2 * math.pi
         return super()._revolve(*args, **kwargs)
 
+    def add_boundary_layer(self, *args, **kwargs):
```
Comment (nschloe) [inline pygmsh/occ/geometry.py:59]: Can't those go in common?

**Comment by nschloe** [comment]
ignore_other_mesh_sizes still needs to be removed.
```
        gmsh.model.mesh.field.setNumber(tag1, "NumPointsPerCurve", self.NumPts_PerCurve)
```
still needs to be removed.

File: pygmsh/occ/geometry.py (hunk 1 above)
Comment (BinWang0213) [reply pygmsh/occ/geometry.py:59]: This is the main purpose of my contribution. Personally, I used this to refine the boundary of a opencascade model

**Comment by BinWang0213** [comment]
> ignore_other_mesh_sizes still needs to be removed.
> 
> ```
>         gmsh.model.mesh.field.setNumber(tag1, "NumPointsPerCurve", self.NumPts_PerCurve)
> ```
> 
> still needs to be removed.

If we use the default value of NumPointsPerCurve=20, the mesh refinement will be wrong for a long edge.
For example:
Short edge:
image
Long edge:
image

Do you have a better idea to change this parameter using the current code?

File: pygmsh/occ/geometry.py (hunk 1 above)
Comment (nschloe) [reply pygmsh/occ/geometry.py:59]: If it's both occ and geo, it can as well go into common. Less code duplication.

Comment (BinWang0213) [reply pygmsh/occ/geometry.py:59]: Good idea. Let me put this into common

File: test/test_boundary_layers.py [hunk 2]
```diff
@@ -35,6 +35,35 @@ def test():
         ref = 4.0
         mesh = geom.generate_mesh()
     assert abs(compute_volume(mesh) - ref) < 1.0e-2 * ref
+
```
Comment (nschloe) [inline test/test_boundary_layers.py:38]: Let's put this in a new function.

File: pygmsh/common/size_field.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            # gmsh.model.mesh.field.setNumbers(tag1, "NodesList", [n._id for n in nodes])
+        if self.faces_list:
+            gmsh.model.mesh.field.setNumbers(
+                tag1, "FacesList", [f._id for f in self.faces_list]
+            )
+        if self.nodes_list:
+            gmsh.model.mesh.field.setNumbers(
+                tag1, "NodesList", [n._id for n in self.nodes_list]
+            )
+        gmsh.model.mesh.field.setNumber(tag1, "NumPointsPerCurve", self.NumPts_PerCurve)
```
Comment (nschloe) [inline pygmsh/common/size_field.py:46]: _Still_ needs to be removed.

File: pygmsh/common/size_field.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(
+        self,
+        lcmin,
+        lcmax,
+        distmin,
+        distmax,
+        edges_list=None,
+        faces_list=None,
+        nodes_list=None,
+        NumPts_PerCurve=100,
```
Comment (nschloe) [inline pygmsh/common/size_field.py:14]: Let's rename this to `num_points_per_curve`.

Comment (nschloe) [reply pygmsh/common/size_field.py:14]: Also, have it default to `None`.

Comment (BinWang0213) [reply pygmsh/common/size_field.py:14]: No problem. Let me change this

**Comment by nschloe** [comment]
Thanks!

**Pull request merged**